numpyro/contrib/module.py

```diff
@@ -566,3 +566,134 @@ def __call__(self, x):
         _update_params(params, new_params, prior)
 
     return partial(apply_fn, new_params, *other_args, **keywords)
+
+
+def eqx_module(name, nn_module):
+    """
+    Declare an :mod:`equinox` style neural network inside a
+    model so that its parameters are registered for optimization via
+    :func:`~numpyro.primitives.param` statements.
+
+    Given an equinox ``nn_module``, to evaluate the module, we directly call it.
+    In a NumPyro model, the pattern will be::
+
+        # Eager initialization outside the model
+        module = nn_module(...)
+
+        # Inside the model
+        net = eqx_module("net", module)
+        y = jax.vmap(net)(x)
+
+    In the case of stateful computation, the pattern is the following::
+
+        # Eager initialization outside the model
+        module, eager_state = eqx.nn.make_with_state(nn_module)(...)
+
+        # Inside the model
+        net = eqx_module("net", module)
+        mutable_holder = numpyro_mutable("net$state", {"state": eager_state})
+        batched_net = jax.vmap(net, in_axes=(0,None), out_axes=(0,None), axis_name='batch')
+        y, state = batched_net(x, mutable_holder['state'])
+        mutable_holder['state'] = jax.lax.stop_gradient(state)
+
+    :param str name: name of the module to be registered.
+    :param eqx.Module nn_module: a pre-initialized `equinox` Module instance.
+    :return: a callable that takes an array as an input and returns
+        the neural network transformed output array.
+    """
+    try:
+        import equinox as eqx
+    except ImportError as e:
+        raise ImportError(
+            "Looking like you want to use equinox to declare "
+            "nn modules. This is an experimental feature. "
+            "You need to install the latest version of `equinox` to use this feature. "
+            "It can be installed with `pip install git+https://github.com/patrick-kidger/equinox.git`."
+        ) from e
+
+    params, static = eqx.partition(nn_module, filter_spec=eqx.is_inexact_array)
+    params = numpyro.param(name + "$params", lambda _: params)
+    nn_module = eqx.combine(params, static)
+    return nn_module
+
+
+def random_eqx_module(name, nn_module, prior):
+    """
+    A primitive to create a random :mod:`equinox` style neural network
+    which can be used in MCMC samplers. The parameters of the neural network
+    will be sampled from ``prior``.
+
+    For supplying a prior dictionary, the dictionary keys are based on their jax key path. In equinox,
+    the key path may often start with a '.' as the first character - this is ignored in the keypath
+    here. To see the jax key paths for all of the leaves in your pytree model, you can run:
+
+        key_paths = [jtu.keystr(path) for path, _ in jtu.tree_leaves_with_path(model_instance)]
```

Comment:
maybe using `keystr(path)[1:]` here and remove the sentence mentioning the dot above.